### ISSUE TITLE
Release v0.16

### ISSUE THREAD

**Issue by marquiz** [kubernetes-sigs/node-feature-discovery#1572]
## Release Checklist

- [ ] All [OWNERS](https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/OWNERS) must LGTM the release proposal
- [x] Verify that the changelog in this issue is up-to-date
- [x] Create new release branch (release v0.$MAJ.0)
  - [x] an OWNER creates a vanilla release branch with
        `git branch release-0.$MAJ master`
  - [x] An OWNER pushes the new release branch with
        `git push release-0.$MAJ`
  - [x] Create Prow pre-submit job configuration for the new release branch in K8s
        [test-infra](https://github.com/kubernetes/test-infra), submit a PR
  - [x] Wait for the test-infra Prow config PR to be merged
- [x] Run `hack/prepare-release.sh $VERSION` to turn references to point to the upcoming release
      (README, deployment templates, docs configuration, test/e2e flags), submit a PR against the release branch
- An OWNER prepares a draft release
  - [x] Create a draft release at [Github releases page](https://github.com/kubernetes-sigs/node-feature-discovery/releases)
  - [x] Write the change log into the draft release
  - [x] Upload release artefacts generated by `prepare-release.sh` script above to the draft release
- [x] An OWNER runs
     `git tag -s -m "NFD release $VERSION" $VERSION`
- [x] An OWNER pushes the tag with
      `git push $VERSION`
  - Triggers prow to build and publish a staging container image
      `gcr.io/k8s-staging-nfd/node-feature-discovery:$VERSION`
  - Triggers build of the documentation and publish it at
        https://kubernetes-sigs.github.io/node-feature-discovery/0.$MAJ/
- [x] Submit a PR against [k8s.io](https://github.com/kubernetes/k8s.io), updating `registry.k8s.io/images/k8s-staging-nfd/images.yaml` to promote the container images (both "full" and "minimal" variants) to production
- [x] Wait for the PR to be merged and verify that the image (`registry.k8s.io/nfd/node-feature-discovery:$VERSION`) is available.
- [x] Publish the draft release prepared at the [Github releases page](https://github.com/kubernetes-sigs/node-feature-discovery/releases)
      which will also trigger a Helm repo index update to add the latest release
- [x] Add a link to the tagged release in this issue.
- [x] Send an announcement email to `dev@kubernetes.io` with the subject `[ANNOUNCE] node-feature-discovery $VERSION is released`
  - [x] Add a link to the release announcement in this issue
- [x] Update README in master branch
  - [x] Update references e.g. by running `hack/prepare-release.sh $VERSION` but **only** committing README.md, and,
        submit a PR
  - [x] Wait for the PR to be merged
- [x] Create an unannotated *devel* tag in the master branch, on the first commit that gets merged after the release branch has been created (presumably the README update commit above), and, push the tag:
      `DEVEL=v0.$(($MAJ+1)).0-devel; git tag $DEVEL master && git push $DEVEL`
      This ensures that the devel builds on the master branch will have a meaningful version number.
- [x] Close this issue

### NodeFeatureGroup API

The NodeFeatureGroup custom resource was added to the NFD API. The NodeFeatureGroup API enables the creation of node groups based on features discovered by NFD. The API is an alpha feature and is disabled by default and can be enabled with the NodeFeatureGroupAPI [feature gate](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/custom-resources.html#nodefeaturegroup).

See [documentation](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/custom-resources.html#nodefeaturegroup) for more details.

### Feature gates

NFD adapted the concept of feature gates from Kubernetes to introduce and stabilize new features in a controlled way. See the [documentation](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/reference/feature-gates.html) for more details. Two existing features ([NodeFeature API](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/reference/feature-gates.html#nodefeatureapi) and [disabling label auto-prefixing](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/reference/feature-gates.html#disableautoprefix)) were converted into feature gates.

#### Upcoming changes

Support for [hooks](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/customization-guide.html#hooks) is deprecated since [v0.12.0](https://github.com/kubernetes-sigs/node-feature-discovery/releases/tag/v0.12.0) and will be completely dropped in the NFD v0.17.

#### RDT feature labels removed

The `feature.node.kubernetes.io/cpu-rdt.*` feature labels that were deprecated in NFD [v0.13](https://github.com/kubernetes-sigs/node-feature-discovery/releases/tag/v0.13.0) were removed. RDT features are still available for use in [NodeFeatureRules](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/custom-resources.html#nodefeaturerule) for custom labels.

#### Deprecated flags and options

The [autoDefaultNs](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/reference/master-configuration-reference.html#autodefaultns) config file option of nfd-master is deprecated and will be removed in NFD v0.17. Superseded by the [DisableAutoPrefix](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/reference/feature-gates.html#disableautoprefix) feature gate (`featureGates.DisableAutoPrefix` Helm parameter).

The `-enable-nodefeature-api` command line flag of nfd-master and nfd-worker and the corresponding `enableNodeFeatureApi` [Helm chart parameter](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/deployment/helm.html#general-parameters) have been deprecated and will be removed in NFD v0.17. Superseded by the [NodeFeature API](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/reference/feature-gates.html#nodefeatureapi) feature gate (`featureGates.NodeFeatureAPI` Helm parameter).

The `-crd-controller` command line flag of nfd-master is deprecated and will be removed with the gRPC API in a future release.

#### Network devices

Discover speed of virtual network interfaces.

#### DMI

Added support for detecting DMI attributes from `/sys/devices/virtual/dmi/id/`. In v0.16 only `sys_vendor` discovered, available as `system.dmiid.sys_vendor` feature for use in [NodeFeatureRules](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/custom-resources.html#nodefeaturerule).

#### Swap

Discover the availability of swap on the node. Available as `memory.swap.enabled` feature for use in [NodeFeatureRules](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/custom-resources.html#nodefeaturerule).

#### Helm chart

Now all nodes are cleaned up (feature labels, annotations, extended resources and taints are removed) after uninstalling NFD using a post-delete hook.

The Helm chart now sets resource requests (cpu and memory) for NFD pods. Users may want to adjust these for their cluster. An option to set the pod priority class was added. See [Helm chart parameters](https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/deployment/helm.html#chart-parameters) in the documentation).

#### Container health

A gRPC health server was added to the nfd-master, nfd-worker and nfd-topology-updater daemons. Deployments (Helm and kustomize) configure container liveness and readiness probes to use that for health checking.

## List of PRs

- github: update tagging instructions in release checklists (#1527)
- Update readme to v0.15.0 release (#1524)
- makefile: fix build: target (#1528)
- Makefile: add -timeout argument to e2e-tests (#1526)
- helm: add post-delete hook that cleans up the node (#1532)
- deployment/kustomize: drop the sample cert-manager overlay (#1534)
- nfd-master: run a separate gRPC health server (#1535)
- source/network: discover speed of virtual network interfaces (#1536)
- go.mod: update dependencies (#1539)
- chore: combine cpu count and thread_siblings functions into discover topology function (#1505)
- source/cpu: drop deprecated cpu-rdt labels (#1530)
- Update readme to v0.15.1 release (#1552)
- hack/generate: patch auto-generated deepcopy functions (#1553)
- apis/nfd: Trivial typo fix in tests (#1537)
- docs: update docs build dependencies (#1543)
- topology-updater: initialize properly with -no-publish (#1554)
- topology-updater: document the -no-publish flag correctly (#1555)
- Wrap nested errors (#1558)
- Prevent `nfd-worker` erroring when reading attributes from paravirtual devices (#1557)
- pkg/utils: move GetKubeconfig from pkg/apihelper here (#1562)
- OWNERS: add AhmedGrati as a reviewer (#1564)
- deployment/helm: don't deploy topology-updater conf unnecessarily (#1565)
- topology-updater: get topology api client directly (#1566)
- pkg/utils: move JsonPatch from pkg/apihelper (#1568)
- nfd-master: ditch apihelper (#1570)
- topology-updater: ditch apihelper (#1567)
- Drop pkg/apihelper (#1561)
- nfd-master: fix node status patching (#1571)
- nfd-topology-updater add pods fingerprint by default (#1560)
- docs: add KEP of Spiffe integration (#1444)
- docs: document removal of hooks in v0.17 (#1573)
- build(deps): bump github.com/opencontainers/runc from 1.1.10 to 1.1.12 (#1575)
- build(deps-dev): bump nokogiri from 1.16.0 to 1.16.2 in /docs (#1576)
- scripts/test-infra: bump golangci-lint to v1.56.1 (#1580)
- scripts/test-infra: bump k8s logcheck to v0.8.1 (#1583)
- Bump Go to v1.22 (#1579)
- scripts/test-infra: bump helm to v3.14.0 (#1582)
- source/kernel: add unit tests for kernel version parsing (#1588)
- helm: add priorityClassName option (#1587)
- source/pci: add unit test for the pci source (#1589)
- nfd-master: log errors on node update retries (#1591)
- source/system: Add reading vendor information (#1574)
- source/cpu: fix build tags on rdt discovery (#1594)
- helm: add ability to use a custom issuer (#1598)
- fix hook issue (#1604)
- generate: update autogenerate tools (#1606)
- apis/nfd/validate: use testify/assert for checking test results (#1590)
- Update readme to v0.15.2 release (#1611)
- Update generate scripts to use latest code_gen functions (#1605)
- nfd-master: mark the -crd-controller flag as deprecated (#1612)
- build(deps): bump google.golang.org/protobuf from 1.32.0 to 1.33.0 (#1613)
- Use close to signal stop channedl in worker and topology-updater (#1620)
- nfd-master: fix memory leak in nfd api-controller (#1615)
- Update readme to v0.15.3 release (#1628)
- Add FeatureGate framework to handle new features (#1623)
- replace AhmedGrati account with TessaIO as reviewer (#1630)
- add swap support in nfd (#1585)
- nfd-master: check if node exists before trying update (#1595)
- Remove references to -enable-nodefeature-api flag (#1632)
- Add owner reference to NRT object (#1602)
- nfd-master: retry node updates indefinitely (#1596)
- nfd-worker: Add liveness probe (#1609)
- topology-updater: Set APIVersion, Kind in the OwnerReference explicitly (#1634)
- helm: fix invalid name of host-swaps volume (#1635)
- nfd-master: do nfd API scheme registration in an init function (#1641)
- chore/deployment: add resources requests and limits for helm and Kustomize (#1631)
- nfd-topology-updater: Add liveness probe (#1643)
- nfd-master: get node object only once when updating node (#1652)
- chore/deploy: make interval property in PodMonitor configurable (#1639)
- nfd-master: protect node updater pool queueing with a lock (#1642)
- nfd-master: prevent crash on empty config struct (#1657)
- Update readme to v0.15.4 release (#1650)
- Tidy up usage of channels for signaling (#1656)
- nfd-master: implement opts for modifying NfdMaster instance (#1658)
- nfd-master: parse kubeconfig even with NoPublish set (#1655)
- Move NFD api to a separate go mod (#1600)
- api/nfd: run go mod tidy (#1661)
- Fix Make generate   (#1662)
- apis/nfd/validate: loosen validation of feature annotations (#1633)
- nfd-master: use separate k8s api clients for each updater (#1653)
- nfd-master: stop node-updater pool before reconfiguring api-controller (#1660)
- build(deps): bump golang.org/x/net from 0.20.0 to 0.23.0 (#1665)
- chore/nfd-master: remove warnings in nfd-master unit tests file (#1668)
- build(deps): bump golang.org/x/net from 0.20.0 to 0.23.0 in api/nfd (#1666)
- apis/nfd: add unit tests for match name functions (#1667)
- apis/nfd: no error on ops that never match (#1670)
- api/nfd: use varargs in the NewInstanceFeatures helper (#1669)
- scripts/test-infra: bump golangci-lint to v1.57.2 (#1674)
- add ARMv7 support (#1659)
- docs: document trade-offs in memory configuration (#1651)
- go.mod: bump kubernetes to v1.30 (#1675)
- cloudbuild.yaml: change machine type to e1-highcpu-32 (#1678)
- test/e2e: stop importing kubernetes test/e2e (#1680)
- hack/init-buildx.sh: fix broken patter matching (#1683)
- Disable armv7 builds  (#1677)
- cloudbuild.yaml: downgrade machine type to e2-highcpu-8 (#1685)
- Update update_codegen.sh for v0.30 version of codegen tools  (#1681)
- Dependabot: Add proper dependabot config file (#1679)
- build(deps): bump azure/setup-helm from 3 to 4 (#1686)
- build(deps): bump actions/checkout from 1 to 4 (#1687)
- build(deps): bump golang.org/x/net from 0.23.0 to 0.24.0 (#1689)
- build(deps): bump sigs.k8s.io/yaml from 1.3.0 to 1.4.0 (#1691)
- build(deps): bump github.com/onsi/gomega from 1.31.0 to 1.33.0 (#1692)
- build(deps): bump github.com/onsi/ginkgo/v2 from 2.15.0 to 2.17.2 (#1690)
- build(deps): bump github.com/jaypipes/ghw from 0.8.1-0.20210827132705-c7224150a17e to 0.12.0 (#1688)
- apis/nfd: increase unit test coverage (#1693)
- build: specify buildx builder name everywhere (#1684)
- source/kernel: silence misleading error on selinux detection (#1694)
- build(deps): bump github.com/klauspost/cpuid/v2 from 2.2.6 to 2.2.7 (#1695)
- build(deps): bump github.com/stretchr/testify from 1.8.4 to 1.9.0 (#1696)
- build(deps): bump github.com/google/uuid from 1.5.0 to 1.6.0 (#1698)
- build(deps): bump github.com/onsi/gomega from 1.33.0 to 1.33.1 (#1699)
- build(deps): bump github.com/k8stopologyawareschedwg/noderesourcetopology-api from 0.1.0 to 0.1.2 (#1697)
- build(deps): bump golang.org/x/net from 0.24.0 to 0.25.0 (#1701)
- build(deps): bump google.golang.org/grpc from 1.60.1 to 1.63.2 (#1702)
- build(deps-dev): bump nokogiri from 1.16.2 to 1.16.5 in /docs (#1706)
- build(deps): bump google.golang.org/protobuf from 1.33.0 to 1.34.1 (#1703)
- build(deps): bump github.com/k8stopologyawareschedwg/podfingerprint from 0.1.2 to 0.2.2 (#1705)
- nfd-master: add DisableAutoPrefix feature gate (#1707)
- Re-add -enable-nodefeature-api cmdline flag (#1708)
- build(deps): bump rexml from 3.2.6 to 3.2.8 in /docs (#1709)
- build(deps): bump github.com/onsi/ginkgo/v2 from 2.17.2 to 2.17.3 (#1711)
- Add NodeFeatureGroup API (#1487)
- api/nfd: document all undocumented fields in the types (#1714)
- nfd-worker: improved log when creating NodeFeature object (#1713)
- apis/nfd: allow different types of features of the same name (#1671)
- cpu: advertise AVX10 version (#1673)
- source/cpu: disable AVX10 label (#1715)
- docs/helm: document all feature gates (#1716)
- build(deps): bump github.com/onsi/ginkgo/v2 from 2.17.3 to 2.19.0 (#1717)
- docs: add more cross-references to NodeFeatureGroup API (#1718)

**Comment by marquiz** [comment]
Created this issue as a placeholder to capture the release note.

**Comment by ArangoGutierrez** [comment]
We are ready to kick off this Release!

**Comment by marquiz** [comment]
GitHub release page:
https://github.com/kubernetes-sigs/node-feature-discovery/releases/tag/v0.16.0

**Comment by marquiz** [comment]
Release announcement:
https://groups.google.com/a/kubernetes.io/g/dev/c/4K9uuWpHe0k
